// react.di/lib/models/Module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var container_service_1 = require("../container-service");
var MODULE_META_DATA = 'react.di-module';
var Module = /** @class */ (function () {
    function Module(target, options) {
        this.target = target;
        this.options = options;
    }
    Module.prototype.getExternalContainer = function (context) {
        var container = context.childContainers.get(this.target);
        if (!container) {
            container = container_service_1.createContainerWithBindings(this.options);
            context.childContainers.set(this.target, container);
        }
        return container;
    };
    Module.prototype.getInternalContainer = function (context) {
        var container = this.getExternalContainer(context);
        if (this.options.imports) {
            container = this.getContainerWithImports(this.options.imports, container, context);
        }
        return container;
    };
    Module.prototype.getContainerWithImports = function (imports, container, context) {
        return imports.reduce(function (container, toImportTarget) {
            var module = Module.getModule(toImportTarget);
            var toImportContainer = module.getExternalContainer(context);
            return container_service_1.merge(container, toImportContainer);
        }, container);
    };
    Module.getModule = function (target) {
        return Reflect.getMetadata(MODULE_META_DATA, target);
    };
    Module.setModule = function (target, moduleMeta) {
        Reflect.defineMetadata(MODULE_META_DATA, moduleMeta, target);
    };
    return Module;
}());
exports.Module = Module;
